### PR TITLE
[16.0][MIG] product_attribute_set

### PR DESCRIPTION
**Pull request by sbejaoui**

### CODE REVIEW

**Review by lmignon — APPROVED** [review]
Thank you @sbejaoui LGTM (Code review + functional tests)

**Comment by GuillemCForgeFlow** [comment]
Hi @sbejaoui, I've added some improvements here: https://github.com/OCA/odoo-pim/pull/125. Could you also add them here?
Thank you 😄

**Review by Camille0907 — APPROVED** [review]
Migration LGTM

File: product_attribute_set/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# Copyright 2015 Akretion (http://www.akretion.com).
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+{
+    "name": "Product Attribute Set",
+    "version": "16.0.1.0.0",
+    "category": "Generic Modules/Others",
```
Comment (simahawk) [inline product_attribute_set/__manifest__.py]: shouldn't we introduce a new category "PIM"?

File: product_attribute_set/models/product.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+from odoo import api, fields, models
+
+
+class ProductTemplate(models.Model):
+    """The mixin 'attribute.set.owner.mixin' override the model's fields_view_get()
+    method which will replace the 'attributes_placeholder' by a group made up of all
+    the product.template's Attributes.
+    Each Attribute will have a conditional invisibility depending on its Attriute Sets.
+    """
```
Comment (simahawk) [inline product_attribute_set/models/product.py]: This explanation belongs to the mixin
```suggestion
```

File: product_attribute_set/models/product.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        default=lambda self: self._get_default_att_set(),
+    )
+
+    def _get_default_att_set(self):
+        """Fill default Product's attribute_set with its Category's
+        default attribute_set."""
+        default_categ_id_id = self._get_default_category_id()
+        if default_categ_id_id:
+            default_categ_id = self.env["product.category"].search(
+                [("id", "=", default_categ_id_id.id)]
```
Comment (simahawk) [inline product_attribute_set/models/product.py]:
```suggestion
                [("id", "=", default_categ_id_id.id)], limit=1
```

File: product_attribute_set/models/product.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+    attribute_set_id = fields.Many2one(
+        "attribute.set",
+        "Attribute Set",
+        default=lambda self: self._get_default_att_set(),
+    )
+
+    def _get_default_att_set(self):
+        """Fill default Product's attribute_set with its Category's
+        default attribute_set."""
```
Comment (simahawk) [inline product_attribute_set/models/product.py]:
```suggestion
        """Get default product's attribute_set by category."""
```

File: product_attribute_set/models/product.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        return super().create(vals_list)
+
+    def write(self, vals):
+        if not vals.get("attribute_set_id") and vals.get("categ_id"):
+            category = self.env["product.category"].browse(vals["categ_id"])
+            vals["attribute_set_id"] = category.attribute_set_id.id
+        return super().write(vals)
+
+    @api.onchange("categ_id")
+    def update_att_set_onchange_categ_id(self):
```
Comment (simahawk) [inline product_attribute_set/models/product.py]:
```suggestion
    def onchange_categ_id(self):
```

File: product_attribute_set/models/product.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        self.ensure_one()
+        if self.categ_id and not self.attribute_set_id:
+            self.attribute_set_id = self.categ_id.attribute_set_id
+
+
+# TODO : add the 'attribute.set.owner.mixin' to product.product in order to display
```
Comment (simahawk) [inline product_attribute_set/models/product.py:51]: TODO still valid?

File: product_attribute_set/models/product_category.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    _inherit = "product.category"
+
+    attribute_set_id = fields.Many2one(
+        "attribute.set",
+        "Default Attribute Set",
+        context={"default_model_id": "product.template"},
+    )
+
+    def write(self, vals):
+        """Fill Category's products with Category's default attribute_set_id if empty"""
```
Comment (simahawk) [inline product_attribute_set/models/product_category.py]:
```suggestion
        # Fill Category's products with Category's default attribute_set_id if empty
```

File: product_attribute_set/models/product_category.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+        super().write(vals)
+        for record in self:
+            if vals.get("attribute_set_id"):
+                product_ids = self.env["product.template"].search(
+                    [("categ_id", "=", record.id), ("attribute_set_id", "=", False)]
+                )
+                for product_id in product_ids:
+                    product_id.attribute_set_id = record.attribute_set_id
```
Comment (simahawk) [inline product_attribute_set/models/product_category.py]:
```suggestion
                templates = self.env["product.template"].search(
                    [("categ_id", "=", record.id), ("attribute_set_id", "=", False)]
                )
                templates.product_id.attribute_set_id = record.attribute_set_id
```

**Comment by simahawk** [comment]
@sbejaoui just a reminder:  ` - `  in the commit msg are not needed :wink:

**Comment by victorvermot** [comment]
@sbejaoui Could you attend to simahawk's comments and rebase PR ? Ty in advance

File: test-requirements.txt [hunk 9]
```diff
@@ -0,0 +1 @@
+odoo-addon-attribute-set @ git+https://github.com/OCA/odoo-pim.git@refs/pull/141/head#subdirectory=setup/attribute_set
```
Comment (santostelmo) [inline test-requirements.txt]: As the attribute_set module is merged we don't need that requirement anymore.

File: product_attribute_set/models/product.py (hunk 6 above)
Comment (simahawk) [reply product_attribute_set/models/product.py:51]: ping :)

File: product_attribute_set/models/product_category.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+        "attribute.set",
+        "Default Attribute Set",
+        context={"default_model_id": "product.template"},
+    )
+
+    def write(self, vals):
+        # Fill Category's products with Category's default attribute_set_id if empty
+        super().write(vals)
+        for record in self:
+            if vals.get("attribute_set_id"):
```
Comment (simahawk) [inline product_attribute_set/models/product_category.py:21]: can we have any test here?

**Review by simahawk — APPROVED** [review]
LG overall. Cleaning up commits as per #141  would be nice ;)

**Comment by rousseldenis** [comment]
I've added a dependency on #171 to remove domain from view (as it causes access rights problems for normal users)

**Comment by lmignon** [comment]
@rousseldenis  https://github.com/OCA/odoo-pim/pull/171 is merged.. This one is ready to be merged

**Pull request merged**